File: image-sync-worker/src/lib/checksum.ts
import { createHash } from 'node:crypto';
import type { Readable } from 'node:stream';
import { logger } from './logger.js';

export async function computeChecksum(
  stream: Readable,
  algorithm: 'md5' | 'sha256' = 'md5',
): Promise<string> {
  return new Promise((resolve, reject) => {
    const hash = createHash(algorithm);
    let streamError: Error | null = null;

    stream.on('error', (err) => {
      streamError = err;
      reject(err);
    });

    stream.on('data', (chunk: Buffer) => {
      hash.update(chunk);
    });

    stream.on('end', () => {
      if (streamError) return;
      const digest = hash.digest('hex');
      logger.debug({ algorithm, digest: digest.slice(0, 16) }, 'Checksum computed');
      resolve(digest);
    });
  });
}

export async function computeFileChecksumFromPath(
  smb: { createReadStream: (path: string) => Readable },
  filePath: string,
  algorithm: 'md5' | 'sha256' = 'md5',
): Promise<string> {
  const stream = smb.createReadStream(filePath);
  return computeChecksum(stream, algorithm);
}

File: image-processing-worker/src/checksum.py
import hashlib
from src.logger import logger


def compute_sha256(filepath: str) -> str:
    hash_sha256 = hashlib.sha256()
    with open(filepath, "rb") as f:
        for chunk in iter(lambda: f.read(65536), b""):
            hash_sha256.update(chunk)

    digest = hash_sha256.hexdigest()
    logger.debug("SHA256 computed", filepath=filepath, digest=digest[:16])
    return digest


def compute_sha256_bytes(data: bytes) -> str:
    digest = hashlib.sha256(data).hexdigest()
    return digest
